Prepare for inclusion in liboqs.

shane-digi opened the pull request:
* Add the META files needed for oqs to consume sqisign.
* Add extra functions to the namespace to prevent collisions.
* Add new functions for sign and verify to follow the oqs conventions.

shane-digi committed:
Add lvl1 yaml file for import to liboqs.

shane-digi committed:
Add levels 3 and 5.

shane-digi committed:
Update namespace for oqs.

shane-digi commented:
I have only tested this on MacOS (M1) at the moment. There are some open questions I have for you.
1. Do you want these changes in this repo or should I/oqs maintain a fork?
2. Is just using mini-gimp for liboqs going to cause performance or license issues?
3. Where would you like the new sign and verify functions to be put (file/header)?

bhess commented on META/sqisign_lvl1.yml:
```diff
@@ -0,0 +1,62 @@
+length-signature: 148
+nistkat-sha256: cce2e6a0e6aff1179eba6b8b5cb1b096e751f9aafdf8934a2df857164a902202
+principal-submitters:
+  - MariusA. Aardal
```
There seem to be some whitespace issues with the names.

shane-digi replied:
I seem to be missing that. Can you give an example?

shane-digi replied:
For the GMP issue, we'll have to figure out how much of a change liboqs wants to make for this. Their only dependency right now is a conditional one on OpenSSL. This would end up being another fixed dependency. Who should I talk to on the oqs team to discuss that?

jordanbray replied:
I believe @bhess is referring to the text "MariusA. Aardal", which should be "Marius A. Aardal"

bhess reviewed (commented):
Hi @shane-digi, thanks for your PR and for working on the liboqs integration!

> 1. Do you want these changes in this repo or should I/oqs maintain a fork?

I’d recommend including the changes in this repo. With the .yml files you’ve added, liboqs can automatically import and update the code as needed.

> 2. Is just using mini-gimp for liboqs going to cause performance or license issues?

Since liboqs avoids including (L)GPL-licensed dependencies, it's preferable not to import mini-gmp. A common alternative is to dynamically link against the system GMP library at runtime (using -DGMP_LIBRARY=SYSTEM -DENABLE_GMP_STATIC=OFF). We might also have to review the DPE dependency to ensure it aligns with liboqs licensing policies.

> 3. Where would you like the new sign and verify functions to be put (file/header)?

The headers should go to [sig.h](https://github.com/SQISign/the-sqisign/blob/main/include/sig.h).

shane-digi committed:
Get Broadwell tests passing for and x86_64 platform.

shane-digi committed:
Add liboqs signature prototypes to sig.h

shane-digi committed:
Fix whitespace around submitter names.

shane-digi committed:
Get Nix Flake from OQS to pass.

shane-digi committed:
Use linux line endings.

jordanbray commented on META/sqisign_lvl1.yml:
```diff
@@ -0,0 +1,63 @@
+principal-submitters:
+  - Marius A. Aardal
+  - Gora Adj
+  - DiegoF.Aranha
```
Here is another name that is messed up.  I assume Tako BorrisFouotsa is also wrong.  There seem to still be a few typos here.

shane-digi committed:
Try to get the authors names right.

shane-digi commented on META/sqisign_lvl1.yml:
```diff
@@ -0,0 +1,63 @@
+  - Luca De Feo
+  - Max Duparc
+  - Jonathan Komada Eriksen
+  - Tako BorisFouotsa
```
I'll get this one with the next update. Struggling mightily with names...

jordanbray replied:
Nice work - I wrote something similar locally (to handle the namespacing), but now I'm using your version.

shane-digi committed:
Remove mem.

shane-digi committed:
Dashed names instead of underscore.

shane-digi committed:
Remove usage of AES and use OQS SHAKE.

bhess committed:
- Update META.yml: removing mini-gmp sources
- Refactor META folder to integrations/liboqs
- Add RADIX macros to gf code
- Update namespace script to include constants & generate non-generic namespace for integration/liboqs

bhess commented:
I've made a few updates to the PR with the following changes:

- Updated the meta.yml files to avoid importing mini-gmp, and instead indicate linking against gmp.
- Improved the namespace generation script: we have a script that auto-generates the sqisign_namespace.h headers. I’ve extended it to also include the constant symbols, syncing with what you introduced in the PR.
- Refactored the META folder to `integration/liboqs`. The goal is to consolidate integration-specific glue code in one place, similar to the structure used in [mlkem-native](https://github.com/pq-code-package/mlkem-native/tree/main/integration/liboqs). I’ve also moved the custom namespace file into this folder, which should eliminate the need for downstream patches in liboqs.

You can view the changes here:
https://github.com/bhess/the-sqisign/commit/39b09acd532c69e3fb1206b4502572479288df92
Please feel free to include this commit in the PR if it’s helpful.

shane-digi committed:
Remove the DPE library.

shane-digi commented:
@bhess As per your comment on the liboqs review I've removed dpe.h.

shane-digi commented:
Is there anything else I should change?

bhess committed:
Update yml files & move code to integration/liboqs

bhess committed:
fix sources in yml

bhess committed:
Add missing fp_num functions

bhess committed:
- Fix C strict & pedantic warnings/errors
- make fips202 calls compatible with liboqs
- Refactor ibz_t to a struct, to avoid pointer-to-array const-type issues.
- Refactor fp_t to avoid pointer-to-array const issues
- Refactor ibq_t to a struct
- Remove unused functions
- Make functions in headers 'static inline'
- remove unused basis argument from post_LLL_basis_treatment
- Precomp scripts: format according to the modified structures

defeo commented:
Thanks you both for your work. Shane's re-implementation of DPE drops support for those few architectures where doubles are 32-bits. I doubt this is a problem, given we could never run signing on 32-bit archs. Do you confirm, @dgazzoni ?

dgazzoni commented:
To be fair I wasn't aware of architectures where doubles could be 32 bits in length. I'm fully in favor of this change.

bhess commented:
I tested the liboqs integration and it works well. However, some CI builds in liboqs use -Wpedantic, which triggered some compiler warnings treated as errors. I refactored parts of the codebase so it now passes these builds. You can see the additional changes on top of the PR here: https://github.com/bhess/the-sqisign/commit/a8884349ee78b0c4da296c9f8ce6f208910d5ee6

shane-digi commented:
Are we waiting on legal for this?

defeo commented:
yes, we're waiting for feedback from lawyers at the Linux Foundation.
